### DISCUSSION TITLE
How to Stop Client Disconnecting

### DISCUSSION THREAD

**Discussion by Lextar** [jchristn/SuperSimpleTcp#78] [Q&A]
Hi, I have a SimpleTCP client which starts the connection and runs for various lengths of time between 5 minutes and 20 minutes and disconnects. 
ClientDisconnectedEventArgs.Reason =  Normal.
I have made sure 
Client.EnableTcpKeepAlives = true;
Client.Keepalive.TcpKeepAliveInterval = 20;
Client.Keepalive.TcpKeepAliveTime = 20;
Client.Keepalive.TcpKeepAliveRetryCount = 10;
Using Visual Studio 2019 up to date and Windows Forms Application .NET 4.7.2 
Is there anything else I can do to maintain the connection?
Thank You

**Comment by jchristn** [comment]
Hi, try disabling keepalives. They can be problematic.

**Reply by Lextar** [reply]
Hi, 
    Thanks for your reply. I did not have them initially but add them when the problem started.
    I will remove specifically disable them and try
Regards

**Reply by Lextar** [reply]
Hi,
    I specifically set Client.Keepalive.EnableTcpKeepAlives = false;
    It disconnected after 5 min 22 sec
Regards

**Reply by jchristn** [reply]
Did you disable on both client side and server side?  

You mentioned you are using a forms app.  Where are you initializing the client and server classes?  Within the form, or in a global context?  Unintentional disconnects are often the result of SimpleTcp getting disposed because it's initialized inside a form class. 

It's really hard to identify why a TCP session was torn down; it could be for any number of reasons.  If you have a packet capture you can see the presence of FIN or RST packets and who sent them.

**Reply by Lextar** [reply]
Thank You, Yes is in a forms app and defined in the main form as static
    public partial class Wagon_Monitor : Form
    {
        static SimpleTcpServer GPS_server;
        static SimpleTcpClient HF_A21_client;
        static SimpleTcpClient Cloud_client;
Could you tell me how should I initialize the client and server classes to stop them disconnecting?
Thank You 
Regards

**Reply by jchristn** [reply]
Hi @Lextar I would initialize in app.cs and make it a static global, i.e. inside of ```Program``` as a ```public static``` or ```internal static```.  Then, you could call it from within your form.
```
    static class Program
    {
        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new ServerForm());
        } 
    }
```

**Reply by Lextar** [reply]
Hi, Thank you for your reply. I have tried to define them in program
    static class Program
    {
        public static SimpleTcpServer GPS_server;
        public static SimpleTcpClient HF_A21_client;
        public static SimpleTcpClient Cloud_client;
        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main()
        {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new ServerForm());
        } 
    }
but getting CS013 name GPS_server does not exist in the current context when the GPS_server; is referenced in the form
Thank You 
Regards

**Reply by jchristn** [reply]
Are you accessing it using ```Program.GPS_server```?  Also ```static class Program``` may need to be ```public static class Program```

**Reply by Lextar** [reply]
Thank You, Working but didn't solve the problem of disconnecting after a period of time usually greater than 5 minutes
Regards

**Reply by jchristn** [reply]
Could you try setting the logger on the client and server and pasting the output?
```
client.Logger = Console.WriteLine;
```
